backend/src/main/java/com/bakdata/conquery/models/events/CBlock.java

```diff
@@ -200,4 +187,122 @@ public boolean isConceptIncluded(int entity, long requiredBits) {
+			TreeChildPrefixIndex.putIndexInto(treeConcept);
+
+			treeConcept.initializeIdCache(stringStore, bucket.getImp());
+		}
+		// No column only possible if we have just one tree element!
+		else if(treeConcept.countElements() == 1){
+			stringStore = null;
+		}
+		else {
+			throw new IllegalStateException(String.format("Cannot build tree over Connector[%s] without Column", connector.getId()));
+		}
+
+
+		final int[][] mostSpecificChildren = new int[bucket.getNumberOfEvents()][];
+
+		Arrays.fill(mostSpecificChildren, ConceptTreeConnector.NOT_CONTAINED);
+
+		final ConceptTreeCache cache = treeConcept.getCache(bucket.getImp());
+
+		final int[] root = treeConcept.getPrefix();
+
+		for (BucketEntry entry : bucket.entries()) {
+			try {
+				final int event = entry.getEvent();
+
+				// Events without values are omitted
+				// Events can also be filtered, allowing a single table to be used by multiple connectors.
+				if (column != null && !bucket.has(event, column)) {
+					mostSpecificChildren[event] = Connector.NOT_CONTAINED;
+					continue;
+				}
+				String stringValue = "";
+				int valueIndex = -1;
+
+				if (stringStore != null) {
+					valueIndex = bucket.getString(event, column);
+					stringValue = stringStore.getElement(valueIndex);
+				}
+
+				// Lazy evaluation of map to avoid allocations if possible.
+				final CalculatedValue<Map<String, Object>> rowMap = new CalculatedValue<>(() -> bucket.calculateMap(event));
+
+
+				if ((connector.getCondition() != null && !connector.getCondition().matches(stringValue, rowMap))) {
+					mostSpecificChildren[event] = Connector.NOT_CONTAINED;
+					continue;
+				}
+
+				ConceptTreeChild child = cache == null
+						? treeConcept.findMostSpecificChild(stringValue, rowMap)
+						: cache.findMostSpecificChild(valueIndex, stringValue, rowMap);
+
+				// All unresolved elements resolve to the root.
+				if (child == null) {
+					mostSpecificChildren[event] = root;
+					continue;
+				}
+
+				// put path into event
+				mostSpecificChildren[event] = child.getPrefix();
+			}
+			catch (ConceptConfigurationException ex) {
+				log.error("Failed to resolve event " + bucket + "-" + entry.getEvent() + " against concept " + treeConcept, ex);
+			}
+		}
+
+		if (cache != null) {
+			log.trace(
+					"Hits: {}, Misses: {}, Hits/Misses: {}, %Hits: {} (Up to now)",
+					cache.getHits(),
+					cache.getMisses(),
+					(double) cache.getHits() / cache.getMisses(),
+					(double) cache.getHits() / (cache.getHits() + cache.getMisses())
+			);
+		}
+
+		return mostSpecificChildren;
+	}
+
+	public static long[] calculateIncludedConcepts(int entities, Bucket bucket, int[][] mostSpecificChildren) {
+		long[] includedConcepts = new long[entities];
+		for (BucketEntry entry : bucket.entries()) {
+			final int[] mostSpecificChild = mostSpecificChildren[entry.getEvent()];
+
+			for (int i = 0; i < mostSpecificChild.length; i++) {
+
+				final long mask = calculateBitMask(i, mostSpecificChild);
```

Comment:
Die CalculateBitMask kannst du auch in ConceptNode verwenden, dann gibst du das der einzelnen konzepte rein.

Reply:
Also ich muss schon noch die andere Methode aus CQConcept verwenden, aber ich habe es jetzt in ConceptNode gekapselt

Reply:
ah, oder meinst du mit deiner Änderung? ich merge gerade erst develop rein